File: lib.py/inputstream.py
#!/usr/bin/env python
import codecs
import re
EOF = None

class XMLInputStream(object):
    def __init__(self, source, tokenizer, encoding=None):
        """XMLInputStream(source, tokenizer, [encoding])

        source can be either a file-object, local filename or a string.
        """

        # List of where new lines occur
        self.newLines = []
        
        # Need a reference to the tokenizer object to deal with the madness
        # that is XML entities
        self.tokenizer = tokenizer

        # The octet stream
        self.rawStream = self.openStream(source)

        self.defaultEncoding = "UTF-8"

        # Detect encoding if no explicit "transport level" encoding is
        # supplied
        if not encoding:
            encoding = self.detectEncoding()
        self.charEncoding = encoding

        # The Unicode string
        uString = self.rawStream.read().decode(self.charEncoding, 'replace')

        # Normalize newlines and null characters
        uString = re.sub('\r\n?', '\n', uString)
        uString = re.sub('\x00', u'\uFFFD', uString)

        # Convert the unicode string into a list to be used as the data stream
        self.dataStream = uString

        self.queue = []

        # Reset position in the list to read from
        self.reset()

    def openStream(self, source):
        """Produces a file object from source. source can be either a file
        object, local filename or a string.
        """

        # Already a file object
        if hasattr(source, 'read'):
            stream = source
        else:
            # Otherwise treat source as a string and convert to a file object
            import cStringIO
            stream = cStringIO.StringIO(str(source))
        return stream

    def detectEncoding(self):
        # XXX This doesn't cover <?xml encoding=""?>
        encoding = self.detectBOM()
        if encoding is None:
            encoding = self.defaultEncoding
        return encoding

    def detectBOM(self):
        """Attempts to detect at BOM at the start of the stream. If
        an encoding can be determined from the BOM return the name of the
        encoding otherwise return None"""

        bomDict = {
            codecs.BOM_UTF8: 'utf-8',
            codecs.BOM_UTF16_LE: 'utf-16-le', codecs.BOM_UTF16_BE: 'utf-16-be',
            codecs.BOM_UTF32_LE: 'utf-32-le', codecs.BOM_UTF32_BE: 'utf-32-be'
        }

        # Go to beginning of file and read in 4 bytes
        self.rawStream.seek(0)
        string = self.rawStream.read(4)

        # Try detecting the BOM using bytes from the string
        encoding = bomDict.get(string[:3])       # UTF-8
        seek = 3
        if not encoding:
            encoding = bomDict.get(string[:2])   # UTF-16
            seek = 2
            if not encoding:
                encoding = bomDict.get(string)   # UTF-32
                seek = 4

        # Set the read position past the BOM if one was found, otherwise
        # set it to the start of the stream
        self.rawStream.seek(encoding and seek or 0)

        return encoding

    def determineNewLines(self):
        # Looks through the stream to find where new lines occur so
        # the position method can tell where it is.
        self.newLines.append(0)
        for i in xrange(len(self.dataStream)):
            if self.dataStream[i] == u"\n":
                self.newLines.append(i)

    def position(self):
        """Returns (line, col) of the current position in the stream."""
        # Generate list of new lines first time around
        if not self.newLines:
            self.determineNewLines()

        line = 0
        tell = self.tell
        for pos in self.newLines:
            if pos < tell:
                line += 1
            else:
                break
        col = tell - self.newLines[line-1] - 1
        return (line, col)

    def reset(self):
        """Resets the position in the stream back to the start."""
        self.tell = 0

    def char(self):
        """Read one character from the stream or queue if available. Return
        EOF when EOF is reached.
        """
        if self.tokenizer.entityCount > 0:
            self.tokenizer.charCount += 1
            if self.tokenizer.charCount > self.tokenizer.entityValueLen:
                self.tokenizer.entityValueLen = 0
                self.tokenizer.charCount = 0
                self.tokenizer.entityCount = 0
        
        if self.queue:
            return self.queue.pop(0)
        else:
            try:
                self.tell += 1
                return self.dataStream[self.tell - 1]
            except:
                return EOF

    def charsUntil(self, characters, opposite=False):
        """Returns a string of characters from the stream up to but not
        including any character in characters or EOF. characters can be any
        container that supports the in method being called on it.
        """
        charStack = [self.char()]

        # First from the queue
        while charStack[-1] and (charStack[-1] in characters) == opposite \
          and self.queue:
            charStack.append(self.queue.pop(0))

        # Then the rest
        while charStack[-1] and (charStack[-1] in characters) == opposite:
            try:
                self.tell += 1
                charStack.append(self.dataStream[self.tell - 1])
            except:
                charStack.append(EOF)

        # Put the character stopped on back to the front of the queue
        # from where it came.
        self.queue.insert(0, charStack.pop())
        return "".join(charStack)

File: lib.rb/inputstream.rb
require 'iconv'
require 'constants'

class XMLInputStream
  attr_reader :queue

  def initialize(source, tokenizer, encoding=nil)
    """XMLInputStream(source, tokenizer, [encoding])

    source can be either a file-object, local filename or a string.
    """

    # List of where new lines occur
    @newLines = []

    # Need a reference to the tokenizer object to deal with the madness
    # that is XML entities
    @tokenizer = tokenizer

    # The octet stream
    @rawStream = openStream(source)

    @defaultEncoding = "UTF-8"

    # Detect encoding if no explicit "transport level" encoding is
    # supplied
    if not encoding
      encoding = detectEncoding()
    end
    @charEncoding = encoding

    # The Unicode string
    uString = @rawStream.read()
    if uString.respond_to? :force_encoding
      uString.force_encoding(@charEncoding)
    elsif @charEncoding.downcase != 'utf-8'
      uString = Iconv.conv('utf-8', @charEncoding, uString)
    end

    # Normalize newlines and null characters
    uString.gsub! /\r\n?/, "\n"
    uString.gsub! /\x00/, "\xef\xbf\xbd" # U+FFFD

    # Convert the unicode string into a list to be used as the data stream
    require 'generator'
    @dataStream = Generator.new(uString.chars)
    def @dataStream.reset; end

    @queue = []

    # Reset position in the list to read from
    reset()
  end

  def openStream(source)
    """Produces a file object from source. source can be either a file
    object, local filename or a string.
    """

    # Already a file object
    if source.respond_to? :read
      stream = source
    else
      # Otherwise treat source as a string and convert to a file object
      require 'stringio'
      stream = StringIO.new(source.to_s)
    end
    return stream
  end

  def detectEncoding
    # XXX This doesn't cover <?xml encoding=""?>
    encoding = detectBOM()
    if encoding == nil
      encoding = @defaultEncoding
    end
    return encoding
  end

  def detectBOM
    """Attempts to detect at BOM at the start of the stream. If
    an encoding can be determined from the BOM return the name of the
    encoding otherwise return nil"""

    bomDict = {
      "\xef\xbb\xbf" => 'utf-8',
      "\xff\xfe" => 'utf-16-le',
      "\xfe\xff" => 'utf-16-be',
      "\xff\xfe\x00\x00" => 'utf-32-le',
      "\x00\x00\xfe\xff" => 'utf-32-be'
    }

    # Go to beginning of file and read in 4 bytes
    @rawStream.seek(0)
    string = @rawStream.read(4)

    # Try detecting the BOM using bytes from the string
    encoding = bomDict[string[0..2]]       # UTF-8
    seek = 3
    if not encoding
      encoding = bomDict[string[0..1]]   # UTF-16
      seek = 2
      if not encoding
        encoding = bomDict[string]   # UTF-32
        seek = 4
      end
    end

    # Set the read position past the BOM if one was found, otherwise
    # set it to the start of the stream
    @rawStream.seek(encoding ? seek : 0)

    return encoding
  end

  def determineNewLines
    # Looks through the stream to find where new lines occur so
    # the position method can tell where it is.
    @newLines.append(0)
    for i in xrange(len(@dataStream))
      if @dataStream[i] == "\n"
        @newLines.append(i)
      end
    end
  end

  def position
    """Returns (line, col) of the current position in the stream."""
    # Generate list of new lines first time around
    if not @newLines
      determineNewLines()
    end

    line = 0
    tell = @tell
    for pos in @newLines
      if pos < tell
        line += 1
      else
        break
      end
    end
    col = tell - @newLines[line-1] - 1
    return [line, col]
  end

  def reset
    """Resets the position in the stream back to the start."""
    @tell = 0
  end

  def char
    """Read one character from the stream or queue if available. Return
    EOF when EOF is reached.
    """
    if @tokenizer.entityCount > 0
      @tokenizer.charCount += 1
      if @tokenizer.charCount > @tokenizer.entityValueLen
        @tokenizer.entityValueLen = 0
        @tokenizer.charCount = 0
        @tokenizer.entityCount = 0
      end
    end

    unless @queue.empty?
      return @queue.shift
    else
      begin
        @tell += 1
        return @dataStream.next
      rescue
        return EOF
      end
    end
  end

  def charsUntil(characters, opposite=false)
    """Returns a string of characters from the stream up to but not
    including any character in characters or EOF. characters can be any
    container that supports the in method being called on it.
    """
    charStack = [char()]

    # First from the queue
    while charStack[-1] and (characters.include? charStack[-1]) == opposite \
      and not @queue.empty?
      charStack << @queue.shift
    end

    # Then the rest
    while charStack[-1] and (characters.include? charStack[-1]) == opposite
      begin
        @tell += 1
        charStack << @dataStream.next
      rescue
        charStack << EOF
      end
    end

    # Put the character stopped on back to the front of the queue
    # from where it came.
    @queue.insert(0, charStack.pop())
    return charStack.join('')
  end
end
